js/modules/k6/grpc/client.go

```diff
@@ -0,0 +1,479 @@
+		}
+
+		dialer := func(ctx context.Context, addr string) (net.Conn, error) {
+			return state.Dialer.DialContext(ctx, "tcp", addr)
+		}
+		opts = append(opts, grpc.WithContextDialer(dialer))
+
+		ctx, cancel := context.WithTimeout(*ctxPtr, timeout)
+		defer cancel()
+
+		var err error
+		c.conn, err = grpc.DialContext(ctx, addr, opts...)
+		if err != nil {
+			errc <- err
+
+			return
+		}
+		close(errc)
+	}()
+
+	if err := <-errc; err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// InvokeRPC creates and calls a unary RPC by fully qualified method name
+//nolint: funlen,gocognit
+func (c *Client) InvokeRPC(ctxPtr *context.Context,
+	method string, req goja.Value, params map[string]interface{}) (*Response, error) {
+	ctx := *ctxPtr
+	rt := common.GetRuntime(ctx)
+	state := lib.GetState(ctx)
+	if state == nil {
+		return nil, errInvokeRPCInInitContext
+	}
+
+	if c.conn == nil {
+		return nil, errors.New("no gRPC connection, you must call connect first")
+	}
+
+	method = strings.TrimPrefix(method, "/")
+	md := c.mds[method]
+
+	if md == nil {
+		return nil, fmt.Errorf("method %q not found in file descriptors", method)
+	}
+
+	c.tags = state.CloneTags()
+	timeout := 60 * time.Second
+
+	ctx = metadata.NewOutgoingContext(ctx, metadata.New(nil))
+	for k, v := range params {
+		switch k {
+		case "headers":
+			rawHeaders, ok := v.(map[string]interface{})
+			if !ok {
+				continue
+			}
+			for hk, kv := range rawHeaders {
+				// TODO(rogchap): Should we manage a string slice?
+				strVal, ok := kv.(string)
+				if !ok {
+					continue
+				}
+				ctx = metadata.AppendToOutgoingContext(ctx, hk, strVal)
+			}
+		case "tags":
+			rawTags, ok := v.(map[string]interface{})
+			if !ok {
+				continue
+			}
+			for tk, tv := range rawTags {
+				strVal, ok := tv.(string)
+				if !ok {
+					continue
+				}
+				c.tags[tk] = strVal
+			}
+		case "timeout":
+			if t, ok := v.(float64); ok && t > 0.0 {
+				timeout = time.Duration(t) * time.Millisecond
+			}
+		}
+	}
+	if state.Options.SystemTags.Has(stats.TagURL) {
+		c.tags["url"] = fmt.Sprintf("%s/%s", c.conn.Target(), method)
+	}
+
+	parts := strings.Split(method, "/")
+	if state.Options.SystemTags.Has(stats.TagService) {
+		c.tags["service"] = parts[0]
+	}
+	if state.Options.SystemTags.Has(stats.TagMethod) {
+		c.tags["method"] = parts[1]
+	}
+
+	if state.Options.SystemTags.Has(stats.TagRPCType) {
+		// (rogchap) This method only supports unary RPCs
+		// if this is refactored to support streaming then this should
+		// be updated to be based on the method descriptor (IsClientStreaming/IsServerStreaming)
+		c.tags["rpc_type"] = "unary"
+	}
+
+	// Only set the name system tag if the user didn't explicitly set it beforehand
+	if _, ok := c.tags["name"]; !ok && state.Options.SystemTags.Has(stats.TagName) {
+		c.tags["name"] = method
+	}
+
+	reqdm := dynamic.NewMessage(md.GetInputType())
+	b, _ := req.ToObject(rt).MarshalJSON()
+	_ = reqdm.UnmarshalJSON(b)
```

Comment:
I can do the error check; but in your example it would still not fail on bogus data; this would just be unmarshaled into unknown fields. This is a feature of protocol buffers: forward and backward compatibility. I don't think we should error on unknown fields because of this.
This error is almost always guaranteed to be nil as you would get a goja/js syntax error before you would get to this point.

Reply:
Hmm fair enough... In Go, I'm used to working with the generated type-safe gRPC code, so having this ad-hoc untyped JS RPC invocation is a bit scary for me :sweat_smile: I'm worried that a lot of users will shoot themselves in the foot by having a typo in some option name and not noticing... Is there no way to definitively check if what is being sent doesn't respect the protobuf definition and won't be accepted by the service?

Reply:
You can validate what is being sent to the server with the `--http-debug` flag.